**1) Import required libraries**

In [98]:
import pandas as pd
import sqlalchemy
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from scipy.cluster.hierarchy import linkage, dendrogram

**2) Read SQL data into pandas dataframe**

In [2]:
engine = sqlalchemy.create_engine("postgresql+psycopg2://postgres:Ganesh1912@localhost:5433/Speech_Based_Channel_Quality_Classification_Project")
conn = engine.connect()

In [17]:
query1 = "SELECT * FROM speech_data_with_outliers"
df_with_outliers = pd.read_sql(query1, engine)
df_with_outliers.head()
df_with_outliers.shape

(15000, 32)

In [19]:
query2 = "SELECT * FROM speech_data_without_outliers"
df_without_outliers = pd.read_sql(query2, engine)
df_without_outliers.head()
# df_without_outliers.shape

,snr,mfcc_1,mfcc_2,mfcc_3,mfcc_4,mfcc_5,mfcc_6,mfcc_7,mfcc_8,mfcc_9,...,delta_mfcc_9,delta_mfcc_10,delta_mfcc_11,delta_mfcc_12,delta_mfcc_13,zcr,centroid,flatness,ste,channel_quality
0,20,-184.554540,64.731300,29.096723,2.926109,-6.665120,-2.904903,4.056716,13.027200,5.751308,...,-0.689299,-0.036962,-0.503716,0.214528,0.169405,0.151764,2504.149150,0.095547,13.342673,good
1,10,-114.594562,32.743967,18.909236,5.124353,-1.339533,-1.749376,3.326869,7.508756,5.972961,...,-0.374092,-0.225826,-0.266470,-0.002571,0.102867,0.295197,3243.057712,0.294671,14.477806,medium
2,5,-70.577885,20.517402,13.138192,4.261370,-0.708041,-2.189525,0.965365,4.702858,5.788497,...,-0.364638,-0.071598,-0.049541,0.237220,0.141327,0.356491,3518.415960,0.396006,17.081917,poor
3,20,-261.703220,34.369880,-1.104398,-12.305087,-8.957579,-9.511588,-13.597694,-1.988079,1.357070,...,-0.087909,0.217547,0.196671,0.140958,-0.194080,0.309723,3002.814796,0.305691,4.954632,good
4,10,-172.815252,18.926842,2.647982,-7.967069,-7.690994,-5.289009,-6.552969,-1.125174,0.879828,...,-0.117741,-0.194506,-0.074500,0.046154,-0.005903,0.359039,3426.704436,0.354011,5.328989,medium


**3) Prepare X and Y variables**

In [63]:
X1 = df_with_outliers[['mfcc_1','mfcc_2','mfcc_3','mfcc_4','mfcc_5','mfcc_6','mfcc_7','mfcc_8','mfcc_9','mfcc_10','mfcc_11','mfcc_12','mfcc_13',
                       'delta_mfcc_1','delta_mfcc_2','delta_mfcc_3','delta_mfcc_4','delta_mfcc_5','delta_mfcc_6','delta_mfcc_7','delta_mfcc_8',
                      'delta_mfcc_9','delta_mfcc_10','delta_mfcc_11','delta_mfcc_12','delta_mfcc_13',
                       'zcr','centroid','flatness','ste']] ## Model1, X1 with outliers
y1 = df_with_outliers['channel_quality']
X1.head()
X1.shape

(15000, 30)

In [64]:
X2 = df_without_outliers[['mfcc_1','mfcc_2','mfcc_3','mfcc_4','mfcc_5','mfcc_6','mfcc_7','mfcc_8','mfcc_9','mfcc_10','mfcc_11','mfcc_12','mfcc_13',
                       'delta_mfcc_1','delta_mfcc_2','delta_mfcc_3','delta_mfcc_4','delta_mfcc_5','delta_mfcc_6','delta_mfcc_7','delta_mfcc_8',
                      'delta_mfcc_9','delta_mfcc_10','delta_mfcc_11','delta_mfcc_12','delta_mfcc_13',
                       'zcr','centroid','flatness','ste']] ## Model2, X2 without outliers
y2 = df_without_outliers['channel_quality']
X2.head()
X2.shape

(14386, 30)

**4) Split data into train and test data**

In [71]:
## Model 1 with outliers
X1_train, X1_test, y1_train, y1_test = train_test_split(X1, y1, test_size = 0.2, random_state = 42, stratify=y1)
X1_train.shape, X1_test.shape
# X1_train.head()

((12000, 30), (3000, 30))

In [73]:
## Model 2 without outliers
X2_train, X2_test, y2_train, y2_test = train_test_split(X2, y2, test_size = 0.2, random_state = 42, stratify=y2)
X2_train.shape, X2_test.shape
# X2_train.head()

((11508, 30), (2878, 30))

**5) Scale data**

In [74]:
## Model 1 with outliers
Scaler1 = StandardScaler()
X1_train_scaled = Scaler1.fit_transform(X1_train)
X1_test_scaled = Scaler1.transform(X1_test)

le1 = LabelEncoder()
y1_train_enc = le1.fit_transform(y1_train)
y1_test_enc = le1.transform(y1_test)

## Model 2 without outliers
Scaler2 = StandardScaler()
X2_train_scaled = Scaler2.fit_transform(X2_train)
X2_test_scaled = Scaler2.transform(X2_test)

le2 = LabelEncoder()
y2_train_enc = le2.fit_transform(y2_train)
y2_test_enc = le2.transform(y2_test)

**6) Evaluate function to check performance of model**

In [86]:
def evaluate_classification(y_test, y_predict):
    acc = accuracy_score(y_test, y_predict)
    prec = precision_score(y_test, y_predict, average = 'weighted')
    rec = recall_score(y_test, y_predict, average = 'weighted')
    f1 = f1_score(y_test, y_predict, average = 'weighted')
    return acc, prec, rec, f1

**7) Model training and classification**

In [87]:
models = {"LogisticRegression":LogisticRegression(),
          "SVM":SVC(kernel='rbf', C=1, gamma='scale')}

# for model in models.values():
for name, model in models.items():
    model = model
    # print(model)

    ## Model 1 with SNR - model training and prediction
    model1 = model
    model1.fit(X1_train_scaled, y1_train_enc)
    y1_predict = model1.predict(X1_test_scaled)

    ## Model 1 with SNR - performance check
    model1_acc, model1_prec, model1_rec, model1_f1 = evaluate_classification(y1_test_enc, y1_predict)

    print(f"Model1 (with outliers): {name} Performance: ")
    print(f"- Accuracy score is: {model1_acc:.6f}")
    print(f"- Precision score is: {model1_prec:.6f}")
    print(f"- Recall score is: {model1_rec:.6f}")
    print(f"- F1 score is: {model1_f1:.6f}")
    print("\n")

    ## Model 2 without SNR - model training and prediction
    model2 = model
    model2.fit(X2_train_scaled, y2_train_enc)
    y2_predict = model2.predict(X2_test_scaled)

    ## Model 2 without SNR - performance check
    model2_acc, model2_prec, model2_rec, model2_f1 = evaluate_classification(y2_test_enc, y2_predict)

    print(f"Model2 (without outliers): {name} Performance: ")
    print(f"- Accuracy score is: {model2_acc:.6f}")
    print(f"- Precision score is: {model2_prec:.6f}")
    print(f"- Recall score is: {model2_rec:.6f}")
    print(f"- F1 score is: {model2_f1:.6f}")
    print("\n")

Model1 (with outliers): LogisticRegression Performance: 
- Accuracy score is: 0.853667
- Precision score is: 0.853271
- Recall score is: 0.853667
- F1 score is: 0.853189


Model2 (without outliers): LogisticRegression Performance: 
- Accuracy score is: 0.871786
- Precision score is: 0.871662
- Recall score is: 0.871786
- F1 score is: 0.871636


Model1 (with outliers): SVM Performance: 
- Accuracy score is: 0.903000
- Precision score is: 0.902379
- Recall score is: 0.903000
- F1 score is: 0.902586


Model2 (without outliers): SVM Performance: 
- Accuracy score is: 0.921821
- Precision score is: 0.921497
- Recall score is: 0.921821
- F1 score is: 0.921507




**8) Model Performance Summary:**

    - Two models (Logistic Regression and SVM) were trained on speech features's datasets with and without outliers.
    - Results show that:
        - Removing outliers improved performance across both models.
        - SVM outperformed Logistic Regression, by achieving ~92% accuracy.